docs: rework TS guide (#178)
apify/apify-ts: docs/guides/typescript_project.mdx

description: Stricter, safer, and better development experience
---

> Crawlee is TypeScript native library, it provides the type definition directly in the package. This allows writing code with auto-completion for TypeScript and JavaScript code alike. Besides that, projects written in TypeScript can take advantage of compile-time type-checking and avoid many coding mistakes, while providing documentation for functions, parameters and return values. It will also help with refactoring a lot.

Reviewer: ```suggestion
Crawlee is built with TypeScript, which means it provides the type definition directly in the package. This allows writing code with auto-completion for TypeScript and JavaScript code alike. Besides that, projects written in TypeScript can take advantage of compile-time type-checking and avoid many coding mistakes, while providing documentation for functions, parameters and return values. It will also help with refactoring a lot, and ensuring the least amount of bugs will sneak through.
```

## Setting up a TypeScript project

To use TypeScript in our projects, we'll need the following prerequisites:

1. TypeScript compiler `tsc` installed somewhere:

    ```shell script
    npm install --dev typescript
    ```

    TypeScript can be a development dependency in our project, as shown above. There's no need to pollute the production environment or the system's global repository with TypeScript.

> 2. A build script invoking `tsc` and a correctly specified `main` entry point defined in the `package.json`:

Reviewer: ```suggestion
2. A build script invoking `tsc` and a correctly specified `main` entry point defined in the `package.json` (pointing to the built code):
```

    ```json
    {
        "scripts": {
            "build": "tsc"
        },

>         "main": "build/main.js"

Reviewer: Seems like the rest of the guide uses dist

```suggestion
        "main": "dist/main.js"
```

    npm install --dev @types/node
    ```

4. TypeScript configuration file allowing `tsc` to understand the project layout and the features used in the project:

    > We are extending the [`@apify/tsconfig`](https://github.com/apify/apify-tsconfig), it contains [the set of rules](https://github.com/apify/apify-tsconfig/blob/main/tsconfig.json) we believe are worth following.

    > To be able to use feature called [Top level await](https://blog.saeloun.com/2021/11/25/ecmascript-top-level-await.html), we will need to set the `module` and `target` compiler options to `ES2022` or above. This will make the project compile to [ECMAScript Modules](https://nodejs.org/api/esm.html).

    ```json title="tsconfig.json"
    {
        "extends": "@apify/tsconfig",
        "compilerOptions": {
            "module": "ES2022",
            "target": "ES2022",
            "outDir": "dist",

>             "lib": ["DOM"]

Reviewer: Do we want to document the usage of the DOM types too? It's *technically* correct for browser scrapers, but not exactly needed for cheerio for example

```dockerfile title="Dockerfile"
# using multistage build, as we need dev deps to build the TS source code
FROM apify/actor-node:16 AS builder

# copy all files, install all dependencies (including dev deps) and build the project
COPY . ./
RUN npm install --include=dev \
    && npm run build

# create final image
FROM apify/actor-node:16
# copy only necessary files
COPY --from=builder /usr/src/app/package*.json ./
COPY --from=builder /usr/src/app/dist ./dist

# install only prod deps
RUN npm --quiet set progress=false \
    && npm install --only=prod --no-optional

# run compiled code
CMD npm run start:prod
```

### Putting it all together

Let's wrap it up to. In addition to the scripts we described above, we also need to set the `type: 'module'` in the `package.json` to be able to use the Top level await described above. For convenience, we will have 3 `start` scripts, the default one will be an alias to `start:dev`, which is our `ts-node` script that does not require compilation (nor type checking). The production script (`start:prod`) is then used in the `Dockerfile`, after explicit `npm run build` call.

```json title="package.json"
{
    "name": "my-crawlee-project",

>     "type": "module",
>     "dependencies": {

Reviewer: ```suggestion
    "type": "module",
    "main": "dist/main.js",
    "dependencies": {
```
